### PR TITLE
[Dev] Tools utils tests

### PR DESCRIPTION
**Pull request by gohabereg**

### CODE REVIEW

File: test/cypress/tests/modules/Tools.spec.ts [hunk 1]
```diff
@@ -0,0 +1,198 @@
...
+import Tools from '../../../../src/components/modules/tools';
+import {EditorConfig} from "../../../../types";
+import BlockTool from '../../../../src/components/tools/block';
+
+describe('Tools module', () => {
+  const defaultConfig = {
+    tools: {},
+  };
+
+  function constructModule(config: EditorConfig = defaultConfig): Tools {
```
Comment (neSpecc) [inline test/cypress/tests/modules/Tools.spec.ts:17]: add a jsdoc please

File: test/cypress/tests/modules/Tools.spec.ts [hunk 2]
```diff
@@ -0,0 +1,198 @@
...
+    const module = new Tools({
+      config,
+      eventsDispatcher: {},
+    } as any);
+
+    const APIMethods = {
+      method(): void {},
+    };
+
+    module.state = {
```
Comment (neSpecc) [inline test/cypress/tests/modules/Tools.spec.ts:31]: what is the module's `state`?

File: src/components/modules/tools.ts [hunk 3]
```diff
@@ -248,6 +206,48 @@ export default class Tools extends Module {
...
 
+  /**
+   * Returns internal tools
+   * Includes Bold, Italic, Link and Paragraph
+   */
+  private get internalTools(): { [toolName: string]: ToolConstructable | ToolSettings & { isInternal?: boolean } } {
+    return {
+      bold: {
+        class: BoldInlineTool,
+        isInternal: true,
```
Comment (neSpecc) [inline src/components/modules/tools.ts:217]: is it necessary to add this excess param to every item? they are all internal

File: test/cypress/plugins/index.ts [hunk 4]
```diff
@@ -1,5 +1,22 @@
...
+const webpackConfig = require('../../../webpack.config.js');
+const preprocessor = require('@cypress/webpack-preprocessor');
+const codeCoverageTask = require('@cypress/code-coverage/task');
+
+module.exports = (on, config): any => {
+  codeCoverageTask(on, config);
+
+  const options = preprocessor.defaultOptions;
+
+  options.typescript = require.resolve('typescript');
```
Comment (neSpecc) [inline test/cypress/plugins/index.ts:23]: please add a description of what is going on here

File: test/cypress/tests/modules/Tools.spec.ts [hunk 5]
```diff
@@ -0,0 +1,198 @@
...
+          inlineTool: class {
+            public static isInline = true
+
+            public render() {}
+            public surround() {}
+            public checkState() {}
+          } as any,
+          unavailableInlineTool: class {
+            public static isInline = true;
+          } as any,
```
Comment (neSpecc) [inline test/cypress/tests/modules/Tools.spec.ts:112]: does the prepare() with error missed? why it is unavailable

File: test/cypress/tests/tools/BlockTool.spec.ts [hunk 6]
```diff
@@ -0,0 +1,376 @@
+/* tslint:disable:max-classes-per-file */
+import { BlockToolData, ToolSettings } from '../../../../types';
+import { ToolType } from '../../../../src/components/tools/base';
+import BlockTool from '../../../../src/components/tools/block';
+
+describe('BlockTool', () => {
+  const options = {
```
Comment (neSpecc) [inline test/cypress/tests/tools/BlockTool.spec.ts:10]: describe this constant, please. What kind of options?

File: test/cypress/tests/tools/BlockTool.spec.ts [hunk 7]
```diff
@@ -0,0 +1,376 @@
...
+       */
+      constructor({ data, block, readOnly, api, config }) {
+        this.data = data;
+        this.block = block;
+        this.readonly = readOnly;
+        this.api = api;
+        this.config = config;
+      }
+    },
+    config: {
```
Comment (neSpecc) [inline test/cypress/tests/tools/BlockTool.spec.ts:57]: why it still named `config`? I would suggest `settings`

File: test/cypress/tests/tools/BlockTool.spec.ts [hunk 8]
```diff
@@ -0,0 +1,376 @@
...
+
+    expect(tool1.isInternal).to.be.false;
+    expect(tool2.isInternal).to.be.true;
+  });
+
+  context('.settings', () => {
+    it('should return correct value', () => {
+      const tool = new BlockTool(options as any);
+
+      expect(tool.settings).to.be.deep.eq(options.config.config);
```
Comment (neSpecc) [inline test/cypress/tests/tools/BlockTool.spec.ts:121]: settings or config? It's confusing

File: test/cypress/tests/tools/BlockTool.spec.ts [hunk 9]
```diff
@@ -0,0 +1,376 @@
...
+    it('should call Tool reset method', () => {
+      options.constructable.reset = cy.stub();
+      const tool = new BlockTool(options as any);
+
+      tool.reset();
+
+      expect(options.constructable.reset).to.be.calledOnce;
+    });
+
+    it('should not fail if Tool prepare method is not exist', () => {
```
Comment (neSpecc) [inline test/cypress/tests/tools/BlockTool.spec.ts]:
```suggestion
    it('should not fail if Tool reset method is not exist', () => {
```

File: test/cypress/tests/tools/BlockTune.spec.ts [hunk 10]
```diff
@@ -0,0 +1,175 @@
+/* tslint:disable:max-classes-per-file */
+import { ToolSettings } from '../../../../types';
+import { ToolType } from '../../../../src/components/tools/base';
+import BlockTune from '../../../../src/components/tools/tune';
+import { BlockTuneData } from '../../../../types/block-tunes/block-tune-data';
+
+describe('BlockTune', () => {
+  const options = {
```
Comment (neSpecc) [inline test/cypress/tests/tools/BlockTune.spec.ts:11]: describe please

File: test/cypress/tests/tools/BlockTune.spec.ts [hunk 11]
```diff
@@ -0,0 +1,175 @@
...
+       */
+      constructor({ api, settings, block, data }) {
+        this.api = api;
+        this.settings = settings;
+        this.block = block;
+        this.data = data;
+      }
+    },
+    config: {
+      config: {
```
Comment (neSpecc) [inline test/cypress/tests/tools/BlockTune.spec.ts:33]: `config.config` again

File: test/cypress/tests/tools/BlockTune.spec.ts [hunk 12]
```diff
@@ -0,0 +1,175 @@
...
+    it('should call Tool reset method', () => {
+      options.constructable.reset = cy.stub();
+      const tool = new BlockTune(options as any);
+
+      tool.reset();
+
+      expect(options.constructable.reset).to.be.calledOnce;
+    });
+
+    it('should not fail if Tool prepare method is not exist', () => {
```
Comment (neSpecc) [inline test/cypress/tests/tools/BlockTune.spec.ts]:
```suggestion
    it('should not fail if Tool reset method is not exist', () => {
```

File: test/cypress/tests/tools/InlineTool.spec.ts [hunk 13]
```diff
@@ -0,0 +1,193 @@
+/* tslint:disable:max-classes-per-file */
+import { ToolSettings } from '../../../../types';
+import { ToolType } from '../../../../src/components/tools/base';
+import InlineTool from '../../../../src/components/tools/inline';
+
+describe('InlineTool', () => {
+  const options = {
```
Comment (neSpecc) [inline test/cypress/tests/tools/InlineTool.spec.ts:10]: describe pls

File: test/cypress/tests/tools/InlineTool.spec.ts [hunk 14]
```diff
@@ -0,0 +1,193 @@
...
+    it('should call Tool reset method', () => {
+      options.constructable.reset = cy.stub();
+      const tool = new InlineTool(options as any);
+
+      tool.reset();
+
+      expect(options.constructable.reset).to.be.calledOnce;
+    });
+
+    it('should not fail if Tool prepare method is not exist', () => {
```
Comment (neSpecc) [inline test/cypress/tests/tools/InlineTool.spec.ts]:
```suggestion
    it('should not fail if Tool reset method is not exist', () => {
```

**Review by neSpecc — CHANGES_REQUESTED** [review]

File: test/cypress/tests/tools/ToolsCollection.spec.ts [hunk 15]
```diff
@@ -0,0 +1,178 @@
...
+const FakeBlockTune = {
+  ...FakeTool,
+  isTune(): boolean {
+    return true;
+  },
+};
+
+describe('ToolsCollection', (): void => {
+  let collection;
+  const fakeTools = [
```
Comment (neSpecc) [inline test/cypress/tests/tools/ToolsCollection.spec.ts:50]: what is this? Need a description

File: test/cypress/tests/tools/ToolsCollection.spec.ts [hunk 16]
```diff
@@ -0,0 +1,178 @@
+import ToolsCollection from '../../../../src/components/tools/collection';
+import BlockTool from '../../../../src/components/tools/block';
+import InlineTool from '../../../../src/components/tools/inline';
+import BlockTune from '../../../../src/components/tools/tune';
+import BaseTool from '../../../../src/components/tools/base';
+
```
Comment (neSpecc) [inline test/cypress/tests/tools/ToolsCollection.spec.ts:6]: Describe what should be tested here.

File: src/components/modules/tools.ts (hunk 3 above)
Comment (gohabereg) [reply src/components/modules/tools.ts:217]: To know if Tool is internal on the later stage we need to add it here

File: test/cypress/tests/modules/Tools.spec.ts (hunk 5 above)
Comment (gohabereg) [reply test/cypress/tests/modules/Tools.spec.ts:112]: It doesn't have required methods: save, surround, checkState

File: test/cypress/tests/tools/BlockTool.spec.ts (hunk 8 above)
Comment (gohabereg) [reply test/cypress/tests/tools/BlockTool.spec.ts:121]: It was the same before. I'd suggest to discuss and change it separately

**Review by neSpecc — APPROVED** [review]

**Review by khaydarov — APPROVED** [review]

**Pull request merged**